cpp/server/ct-server.cc

```diff
@@ -92,19 +98,39 @@ static const bool cert_dummy =
 
 }  // namespace
 
+void handler(int sig) {
+  void *array[90];
+  size_t size;
+
+  // get void*'s for all entries on the stack
+  size = backtrace(array, 90);
+
+  // print out all the frames to stderr
+  fprintf(stderr, "Error: signal %d:\n", sig);
+  backtrace_symbols_fd(array, size, STDERR_FILENO);
+  exit(1);
+}
 
 int main(int argc, char* argv[]) {
   // Ignore various signals whilst we start up.
   signal(SIGHUP, SIG_IGN);
   signal(SIGINT, SIG_IGN);
   signal(SIGTERM, SIG_IGN);
 
+  signal(SIGSEGV, handler);
+  signal(SIGABRT, handler);
+
   ConfigureSerializerForV1CT();
   util::InitCT(&argc, &argv);
 
   Server::StaticInit();
 
-  util::StatusOr<EVP_PKEY*> pkey(ReadPrivateKey(FLAGS_key));
+  util::StatusOr<EVP_PKEY*> pkey = NULL;
+  if (FLAGS_engine == "") {
```

Comment:
How about renaming `ReadPrivateKey` to `ReadPrivateKeyFromFile`, and breaking this `if/else` out into `ReadPrivateKey()` function in `util/read_key.cc`, then do this here: 
```C++
const util::StatusOr<EVP_PKEY*> key(ReadPrivateKey(FLAGS_key));
```
?